### ISSUE TITLE
jacobians with output_length takes a long time to create

### ISSUE THREAD

**Issue by KristofferC** [JuliaDiff/ForwardDiff.jl#63]
Some background of my use case: 
I need to solve a smallish nonlinear problem in a lot of points where each points have a set of different parameters. For a specific point I make a closure from the parameters of the point and use `ForwardDiff` to compute the jacobian of that closure. I then pass the closure and the jacobian to an external solver and everything works great.

I now wanted to try a solver that only allows for in place updates of the function values. This means I should use the `output_length` option to `jacobian`. When I do that, the analysis performance takes a large hit because from what I believe is the compilation time in generating the function [here](https://github.com/JuliaDiff/ForwardDiff.jl/blob/33f61b3e77afcdcdd20b938e3a6b2def677c6b41/src/api/jacobian.jl#L50) in each new point. 

Since I have a constant `output_length` it should in theory be possible to cache something similar to the `newf` function that  takes the function I want to compute the jacobian of as an extra argument instead of creating the function from scratch.

Would this be possible/sensible?

**Comment by mlubin** [comment]
What's the need for an ``@eval`` on that line?

**Comment by KristofferC** [comment]
Another thing, wouldn't the `output_cache` after the `@eval` be created in global scope and therefore benefit of having a `const` on it?

**Comment by jrevels** [comment]
>Since I have a constant output_length it should in theory be possible to cache something similar to the newf function that takes the function I want to compute the jacobian of as an extra argument instead of creating the function from scratch.

This makes sense, and should definitely be an option in some capacity.

> What's the need for an eval on that line?

It was used so that output_length would be considered a hard-coded literal at the time of compilation. Closing over output_length as a *variable* was causing type instability in the calc_jacobian method, and expanding it as a literal in the expression fixed it (found this out during benchmarking when I originally wrote this). 

I just tried this now on the latest v0.4 release, and it seems like the situation is reversed - type inference now works *without* the eval. 

I'll put in a PR to remove the eval junk 
sometime in the upcoming week (I'm without Internet atm - I'm happy to report that GitHub has nice mobile app).

> Another thing, wouldn't the output_cache after the eval be created in global scope and therefore benefit of having a const on it?

The eval occurs within the scope of the jacobian method, so output_cache is a variable local to the jacobian method.
